lib/hash_ring/ets.ex

```diff
@@ -287,83 +349,184 @@ defmodule ExHashRing.HashRing.ETS do
+    {:ok, do_find_nodes(table, gen, num_nodes, num, Utils.hash(key), [], 0)}
+  end
 
-    config =
-      if map_size(overrides) > 0 do
-        overrides =
-          overrides
-          |> Enum.filter(fn {_, values} -> length(values) > 0 end)
-          |> Map.new()
+  defp do_find_nodes_in_table(key, table, overrides, gen, num_nodes, num) do
+    {found, found_length} =
+      case overrides do
+        %{^key => overrides} ->
+          Utils.take_max(overrides, num)
 
-        {table, new_ring_gen, length(nodes), overrides}
-      else
-        {table, new_ring_gen, length(nodes)}
+        _ ->
+          {[], 0}
       end
 
-    Config.set(name, self(), config)
+    {:ok, do_find_nodes(table, gen, num_nodes, max(num - found_length, 0), Utils.hash(key), found, found_length)}
+  end
+
+  defp do_find_nodes(_table, _gen, _num_nodes, 0, _hash, found, _found_length) do
+    Enum.reverse(found)
+  end
 
-    schedule_ring_gen_gc(%{state | ring_gen: new_ring_gen}, ring_gen)
+  defp do_find_nodes(_table, _gen, num_nodes, _remaining, _hash, found, num_nodes) do
+    Enum.reverse(found)
   end
 
-  def schedule_ring_gen_gc(state, 0), do: state
+  defp do_find_nodes(table, gen, num_nodes, remaining, hash, found, found_length) do
+    {number, node} = find_next_highest_item(table, gen, num_nodes, hash)
+
+    if node in found do
+      do_find_nodes(
+        table,
+        gen,
+        num_nodes,
+        remaining,
+        number,
+        found,
+        found_length
+      )
+    else
+      do_find_nodes(
+        table,
+        gen,
+        num_nodes,
+        remaining - 1,
+        number,
+        [node | found],
+        found_length + 1
+      )
+    end
+  end
+
+  defp update_nodes(%__MODULE__{} = state, nodes) do
+    new_ring_gen = state.ring_gen + 1
+
+    # Get the current generation of items from the current table
+    previous_items =
+      state.current_table
+      |> :ets.match({{state.ring_gen, :"$1"}, :"$2"})
+      |> Enum.map(fn [hash, name] ->
+        {{new_ring_gen, hash}, name}
+      end)
+
+    # Write the previous items into the previous table for this generation
+    :ets.insert(state.previous_table, previous_items)
+
+    # Generate current items for the current table
+    current_items =
+      nodes
+      |> Utils.gen_items()
+      |> Enum.map(fn {hash, name} ->
+        {{new_ring_gen, hash}, name}
+      end)
+
+    # Write the current items into the current table for this generation
+    :ets.insert(state.current_table, current_items)
 
-  def schedule_ring_gen_gc(%{pending_gcs: pending_gcs} = state, ring_gen) do
-    ring_gen_gc_delay =
-      Application.get_env(:hash_ring, :ring_gen_gc_delay, @default_ring_gen_gc_delay)
+    # Update the configuration to atomically cut over to the new generation
+    config = {
+      {state.current_table, length(nodes)},
+      {state.previous_table, length(state.nodes)},
+      new_ring_gen,
+      state.overrides
```

Comment:
What value does config provide us if we are just duplicating most of it's data here in state?

Seems like we could move ownership of this data to config and store a copy into `state.config`? That would allow you to drop the state fields: `current_table, previous_table, ring_gen, overrides`

Reply:
Yea, I'm not really sure why there's duplication between these two components, I could remove some of this.  

This process owns the tables so it seems pretty reasonable they would hang onto the tids, they only get put into the Config I imagine so that you don't have to enter the server context.

Reply:
Well it's doing this so the information will be available in a client available ETS table instead of having to cross back over into either Config's server context or the Ring's server context, because that would incur such a high performance penalty. 

If we have the config just in the internal state of the Ring every query has to single track through the Ring GenServer, which would tank performance.